mlx_graphs/nn/conv/rel_conv.py

```diff
@@ -0,0 +1,327 @@
+            ),
+            aggregate_kwargs=dict(edge_weight=edge_weight, dim_size=size),
+            update_kwargs=dict(input=node_features),
+        )
+        return output
+
+    def message(
+        self,
+        src_features: mx.array,
+        dst_features: mx.array,
+        relation: mx.array,
+        boundary: mx.array,
+        edge_type: mx.array,
+    ) -> mx.array:
+        # if self.input_dims =
+        # extracting relation features
+        relation_j = relation[:, edge_type]
+
+        if self.input_dims == 3:
+            relation_j = relation_j.transpose(1, 0, 2)
+        else:
+            relation_j = relation_j.squeeze(0)
+
+        if self.message_func == "transe":
+            message = src_features + relation_j
+        elif self.message_func == "distmult":
+            message = src_features * relation_j
+        elif self.message_func == "rotate":
+            x_j_re, x_j_im = src_features.chunk(2, dim=-1)
+            r_j_re, r_j_im = relation_j.chunk(2, dim=-1)
+            message_re = x_j_re * r_j_re - x_j_im * r_j_im
+            message_im = x_j_re * r_j_im + x_j_im * r_j_re
+            message = mx.concatenate([message_re, message_im], axis=-1)
+        else:
+            raise ValueError("Unknown message function `%s`" % self.message_func)
+
+        # augment messages with the boundary condition
+        message = mx.concatenate(
+            [message, boundary], axis=0
+        )  # (num_edges + num_nodes, batch_size, input_dim)
+
+        return message
+
+    def aggregate(self, messages, indices, edge_weight, dim_size) -> mx.array:
+        # augment aggregation index with self-loops for the boundary condition
+        index = mx.concatenate(
+            [indices, mx.arange(dim_size[0])]
+        )  # (num_edges + num_nodes,)
+        edge_weight = mx.concatenate([edge_weight, mx.ones(dim_size[0])])
+        shape = [1] * messages.ndim
+        shape[self.node_dim] = -1
+        edge_weight = edge_weight.reshape(shape)
+
+        if self.aggregate_func == "pna":
+            mean = scatter(
+                messages * edge_weight,
+                index,
+                axis=self.node_dim,
+                out_size=dim_size[0],
+                aggr="mean",
+            )
+            sq_mean = scatter(
+                messages**2 * edge_weight,
+                index,
+                axis=self.node_dim,
+                out_size=dim_size[0],
+                aggr="mean",
+            )
+            max = scatter(
+                messages * edge_weight,
+                index,
+                axis=self.node_dim,
+                out_size=dim_size[0],
+                aggr="max",
+            )
+            # scatter_min is not implemented in MLX-graphs
+            # min = scatter(
+            #     messages * edge_weight,
+            #     index, axis=self.node_dim,
+            #     out_size=dim_size[0],
+            #     aggr="min"
+            # )
```

Comment:
Actually, we should implement it - you can take inspiration from https://github.com/mlx-graphs/mlx-graphs/blob/42a58d0855ae496af4b9ab01935c84a314e94549/mlx_graphs/utils/scatter.py#L89 and compute `src.at[index].minimum(values)`

Reply:
added in https://github.com/mlx-graphs/mlx-graphs/pull/111/commits/c7e3779b392dd893ba7b75a36402fa8ba31240fd